text/0068-delayed-refunds.md

```diff
@@ -0,0 +1,104 @@
+# Delayed refunds
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0068](https://github.com/casperlabs/ceps/pull/0068)
+
+We propose enabling the dormant gas refund feature, additionally introducing a variable delay to disbursement of refunded tokens which depends on recently observed user behavior.
+
+## Motivation
+
+[motivation]: #motivation
+
+Due to the inherent uncertainty in gas costs caused by variation in inputs, users almost always need to overclaim gas to eliminate risk of reversion. Because refunds are always set to 0, this creates a serious UX problem. Additionally, a complete resolution to this problem would require full refunds. Unfortunately, given the execution-after-consensus paradigm Casper uses, full refunds create a serious attack vector, enabling malicious or careless users to fill up blocks with overclaimed gas.
+
+To address both the UX issue, and the security issue with refunds, we propose a delay to disbursement of refunded tokens, which would always be held for at least one era. However, depending on the proportion of gas refunded to the user, the delay would grow longer. The length of the delay is governed by the amount of gas recently refunded.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+### System view
+
+With the implementation of this proposal, system contracts would maintain some metadata regarding accounts reflecting their recent usage of gas and penalizing poor gas estimation (high refunds) by delaying the disbursement of tokens due to be refunded. It is expected that the parameters would be set in such a way as to prevent this ever being an issue for a user with gas use estimates off by less than 10%, with such users receiving their entire refund at the end of each era.
+
+### User view
+
+From the user's perspective, the refunds are full but possibly delayed. It is expected the delayed refunds will incentivize use of whichever gas cost estimation tools may be available. Additionally, delayed refunds should disincentivize exploitation of full refunds for malicious purposes, since any attempt to maliciously overclaim gas at low cost would result in prohibitively lengthy refund disbursement schedules.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Refund delay logic
+
+We will denote state variables evolving as eras switch in bold letters.
+
+#### Elements
+
+*Thresholds* in {p<sub>1</sub>, ..., p<sub>N</sub>} are percentages (fractions ranging from 0 to 1) that define ranges of observed past refunds, which determine a *penalty* in {1, ..., M}.
+
+Fix N to be #{1, ..., M}. The thresholds define the penalty in the following way:
+
+- Take p<sub>i</sub> and p<sub>i+1</sub>
+- If one minus *aggregate utilization* **P** (1 - **P**) is between p<sub>i</sub> and p<sub>i+1</sub>
```

Comment:
I read this as "P · (1 - P)", i.e. "P times one minus P" at first. Maybe less ambiguous:

```suggestion
- If 1 - **P**, i.e. one minus *aggregate utilization* **P**, is between p<sub>i</sub> and p<sub>i+1</sub>
```

Reply:
Agreed